src/annoylib.h

```diff
@@ -88,6 +88,60 @@ inline void normalize(T* v, int f) {
     v[z] /= norm;
 }
 
+template<typename S, typename T, typename Random, typename Distance>
+inline void two_means(const vector<typename Distance::Node*>& nodes, int f, Random& random, bool cosine, T* best_iv, T* best_jv) {
+  size_t count = nodes.size();
+
+  T best_d_sum = numeric_limits<T>::infinity();
+  vector<T> iv(f, 0), jv(f, 0), iv_sum(f, 0), jv_sum(f, 0);
+  
+  for (int attempt = 0; attempt < 5; attempt++) {
+    size_t i = random.index(count);
+    size_t j = random.index(count-1);
+    j += (j >= i); // ensure that i != j
+    std::copy(&nodes[i]->v[0], &nodes[i]->v[f], &iv[0]);
+    std::copy(&nodes[j]->v[0], &nodes[j]->v[f], &jv[0]);
+    if (cosine) { normalize(&iv[0], f); normalize(&jv[0], f); }
+
+    for (int iter = 0; iter < 5; iter++) {
+      std::fill(iv_sum.begin(), iv_sum.end(), 0);
+      std::fill(jv_sum.begin(), jv_sum.end(), 0);
+      int ic = 0, jc = 0;
+      T d_sum = 0;
+      for (S l = 0; l < 100 && l < (S)nodes.size(); l++) {
+        size_t k = random.index(count);
+        T di = Distance::distance(&iv[0], nodes[k]->v, f),
+          dj = Distance::distance(&jv[0], nodes[k]->v, f);
+        T norm = cosine ? get_norm(nodes[k]->v, f) : 1.0;
+	d_sum += std::min(di, dj);
+        if (di < dj) {
+          ic++;
+          for (int z = 0; z < f; z++)
+            iv_sum[z] += nodes[k]->v[z] / norm;
+        } else if (dj < di) {
+          jc++;
+          for (int z = 0; z < f; z++)
+          jv_sum[z] += nodes[k]->v[z] / norm;
+        }
+      }
+
+      if (ic == 0 || jc == 0)
+        break;
+
+      if (d_sum < best_d_sum) {
+        best_d_sum = d_sum;
```

Comment:
i think i had it there all the time when i was testing, must have pushed the wrong code for some reason.